sourcemap/fleet_store_test.go

```diff
@@ -77,4 +84,132 @@ func TestFleetFetch(t *testing.T) {
 	assert.True(t, hasAuth)
 }
 
+func TestMultipleFleetHostsQuery(t *testing.T) {
+	var (
+		called        int32
+		apikey        = "supersecret"
+		name          = "webapp"
+		version       = "1.0.0"
+		path          = "/my/path/to/bundle.js.map"
+		c             = http.DefaultClient
+		sourceMapPath = "/api/fleet/artifact"
+	)
+
+	hError := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		atomic.AddInt32(&called, 1)
+		http.Error(w, "err", http.StatusInternalServerError)
+	})
+	ts0 := httptest.NewServer(hError)
+
+	requestReceived := make(chan struct{})
+	h := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		atomic.AddInt32(&called, 1)
+		select {
+		case requestReceived <- struct{}{}:
+			wr := zlib.NewWriter(w)
+			defer wr.Close()
+			wr.Write([]byte(resp))
+		case <-r.Context().Done():
+			return
+		}
+	})
+	ts1 := httptest.NewServer(h)
+	ts2 := httptest.NewServer(h)
+
+	fleetCfg := &config.Fleet{
+		Hosts:        []string{ts0.URL[7:], ts1.URL[7:], ts2.URL[7:]},
+		Protocol:     "http",
+		AccessAPIKey: apikey,
+		TLS:          nil,
+	}
+
+	cfgs := []config.SourceMapMetadata{
+		{
+			ServiceName:    name,
+			ServiceVersion: version,
+			BundleFilepath: path,
+			SourceMapURL:   sourceMapPath,
+		},
+	}
+	f, err := newFleetStore(c, fleetCfg, cfgs)
+	assert.NoError(t, err)
+
+	fetchReturned := make(chan string)
+	go func() {
+		defer close(fetchReturned)
+		resp, _ := f.fetch(context.Background(), name, version, path)
+		fetchReturned <- resp
+	}()
+
+	select {
+	case <-requestReceived:
+	case <-time.After(10 * time.Second):
+		t.Fatal("timed out waiting for server to receive request")
+	}
+
+	select {
+	case resp := <-fetchReturned:
+		assert.Contains(t, resp, "webpack:///bundle.js")
+		assert.EqualValues(t, 3, called)
+	case <-time.After(10 * time.Second):
+		t.Fatal("timed out waiting for Fetch to return")
+	}
+
+	select {
+	case <-requestReceived:
+	default:
+		t.Fatal("No second request received by handler")
+	}
```

Comment:
I think this is maybe a little bit misleading, and IIANM it's racy.

- since we make the requests in parallel, a second request may received by the handler before the request (context) is cancelled
- even though we wait for `<-fetchReturned` above, we don't wait for all requests to complete in the fetch method if we get a success; so it is possible (if unlikely) that the second request will send to `requestReceived`, and this assertion will fail

I'd suggest changing all the above (line 137+) to:

```go
resp, err := f.fetch(context.Background(), name, version, path)
require.NoError(t, err)
assert.Contains(t, resp, "webpack:///bundle.js")
```

Reply:
>even though we wait for <-fetchReturned above, we don't wait for all requests to complete in the fetch method if we get a success

Since the successful handler blocks until a read from `requestReceived`, I read this test as verifying that `fetch` returns after it receives its first successful response, even if there are other requests that haven't returned yet. I think the handler is supposed to receive both requests.

> so it is possible (if unlikely) that the second request will send to requestReceived, and this assertion will fail

I think this is what is being tested, that the second request has sent to `requestReceived`. The successful handler is blocked by a send to `requestReceived`, so a single receive on that and then receive from `requestReceived` means `fetch()` returned after its first successful response, with the second request is still waiting within the handler? And the final receive is verifying that both requests were sent to the handler.

The `ctx` in the handler is not the same as the `ctx` in the request, right? So a successful response + `cancel()` of the request ctx won't effect the `ctx` in the server.

Reply:
I agree that this check is unreliable because the request following the successful request might not reach the handler
```go
select {
	case <-requestReceived:
	default:
		t.Fatal("No second request received by handler")
	}
```

But I still don't understand why this is not relevant for testing that the first time the handler is reached, the fetch method has returned successfully?

```go
select {
    case resp := <-fetchReturned:
        assert.Contains(t, resp, "webpack:///bundle.js")
        assert.EqualValues(t, 3, called)
    case <-time.After(10 * time.Second):
        t.Fatal("timed out waiting for Fetch to return")
    }
```

Reply:
> I think this is what is being tested, that the second request has sent to requestReceived. The successful handler is blocked by a send to requestReceived, so a single receive on that and then receive from requestReceived means fetch() returned after its first successful response, with the second request is still waiting within the handler? And the final receive is verifying that both requests were sent to the handler.
> 
> The ctx in the handler is not the same as the ctx in the request, right? So a successful response + cancel() of the request ctx won't effect the ctx in the server.

@stuartnelson3 right, but the request may never even make it to the 2nd+ servers. Hypothetical ordering:

1. client issues request to ts0, ts1, ts2
2. request is transmitted to ts0
3. ts0 processes request, responds with success
4. client processes response, cancels context
5. because context is cancelled, requests to ts1 and ts2 are never transmitted
6. test fails on second `case <-requestReceived`

It's unlikely, but run it enough times on a slow CI box and it will eventually happen.

> But I still don't understand why this is not relevant for testing that the first time the handler is reached, the fetch method has returned successfully?

@estolfo you're referring to the fetchReturned/timeout bit specifically? The existing `TestFetchContext` test which uses this pattern is doing so because it's testing that asynchronous context cancellation works -- we're not doing that here (at the moment).

I was suggesting the simplification to expedite merging the PR, but I think it would be good to test at this level of detail - it just might take a bit longer. Something like:
 - call fetch in a goroutine (like you're doing)
 - block until all servers have received their request
 - have ts0 respond with failure
 - have ts1 respond with success
 - block until ts2's request is cancelled